src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,170 @@
+
+                        return await IntroduceUsingStatementAsync(context.Document, syntax, symbol, cancellationToken).ConfigureAwait(false);
+                    }));
+            }
+        }
+
+        private static async Task<(TLocalDeclarationSyntax, ILocalSymbol)> FindDisposableLocalDeclaration(Document document, TextSpan selection, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var declarationSyntax = root.FindNode(selection)?.FirstAncestorOrSelf<TLocalDeclarationSyntax>();
+            if (declarationSyntax is null)
+            {
+                return default;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var operation = semanticModel.GetOperation(declarationSyntax, cancellationToken) as IVariableDeclarationGroupOperation;
+            if (operation?.Declarations.Length != 1)
+            {
+                return default;
+            }
+
+            var localDeclaration = operation.Declarations[0];
+            if (localDeclaration.Declarators.Length != 1)
+            {
+                return default;
+            }
+
+            var declarator = localDeclaration.Declarators[0];
+            if (declarator.Initializer is null)
+            {
+                return default;
+            }
+
+            var localType = declarator.Symbol?.Type;
+            if (localType is null)
+            {
+                return default;
+            }
+
+            if (declarator.Initializer.Value?.Syntax is null)
+            {
+                return default;
+            }
+
+            var isValidSelection = await CodeRefactoringHelpers.RefactoringSelectionIsValidAsync(
+                document,
+                selection,
+                node: declarationSyntax,
+                holes: ImmutableArray.Create(declarator.Initializer.Value.Syntax),
+                cancellationToken).ConfigureAwait(false);
+
+            if (!isValidSelection)
+            {
+                return default;
+            }
+
+            var disposableType = semanticModel.Compilation?.GetSpecialType(SpecialType.System_IDisposable);
+            if (disposableType is null || !localType.InheritsFromOrEquals(disposableType, includeInterfaces: true))
+            {
+                return default;
+            }
+
+            return (declarationSyntax, declarator.Symbol);
+        }
+
+        protected static TStatementSyntax FindSiblingStatementContainingLastUsage<TStatementSyntax>(
+            TStatementSyntax declarationSyntax,
+            ILocalSymbol localVariable,
+            SemanticModel semanticModel,
+            ISyntaxFactsService syntaxFactsService,
+            CancellationToken cancellationToken)
+
+            where TStatementSyntax : SyntaxNode
+        {
```

Comment:
an alternative approach that might would would be to get the parent, walk backward to find the last reference.  That itself could be a generic helper elsewhere.  

Then, when you have the node for the last reference, you simply ask for it's ancestor node that shares your parent.  that would be your sibling node.

Reply:
The granularity as at the sibling statement level, so this wouldn't provide any additional performance boost overall. In fact, it would probably go down due to the extra bookkeeping.

Reply:
It was not about a performance boost.  it was about building this out of more general helpers with a simple and clear algorithm.

Reply:
Specifically, all you'd need was a helper to find the last-ref of a local in a specified node.  That seems generally useful and sane to have as a general helper.

Then you could simply use the result of that with exsiting code we have to give the specific answer of "last sibling statement".

Reply:
Okay. I'm not ready to generalize yet. Can I keep this PR more concrete?